### DISCUSSION TITLE
How do i prevent nginx stripping body of request?

### DISCUSSION THREAD

**Discussion by iwwencreative** [serverpod/serverpod#2997] [Q&A]
Hello,
I am facing an issue with my Serverpod deployment in production mode.

Setup Details:
Deployment Mode: Non-Docker deployment, running Serverpod as a systemctl service in production mode.
Reverse Proxy: Using Nginx as a reverse proxy to route requests to the Dart server.
Development Environment: Everything works perfectly in the local development environment.
Endpoints: All endpoints, including adminAuth, are defined and working in development.
Code Generation: I have run serverpod generate before deploying to production.
Issue:
When I try to call the adminAuth endpoint in the production environment, I receive the following error:

DartError: ServerpodClientException: Bad request: No method name specified in call to adminAuth, statusCode = 400
    at Object.throw_ [as throw] (errors.dart:296:3)

Additional Details:
The server is running without any other obvious issues.
Nginx is correctly routing traffic to the Dart server.
I’ve verified that the server is in production mode and that the correct endpoints file is loaded.
Request for Help:
Could this be related to the production setup or the reverse proxy configuration?
Is there something I might have overlooked when deploying Serverpod without Docker in production?
Any debugging steps or insights would be greatly appreciated!

**Comment by iwwencreative** [comment]
in addition here is another error message
No method name specified in call to serverpod_auth.status

**Comment by SandPod** [comment]
Hey @iwwencreative!

### Additional questions
- Does this issue only occur for this endpoint or for all endpoints on the Server?
- Did you ensure all dependencies are up to date in the production environment?
- What version of Serverpod are you running?
- Could you supply the code for how the call is made in the client?
- As far as I know the status endpoint in the authentication module does not provide a method named `adminAuth`. The endpoints provided can be viewed here if you search for `serverpod_auth.status`: https://github.com/serverpod/serverpod/blob/4af8a5c85954fe59fba00b00c1939215bdca6acb/modules/serverpod_auth/serverpod_auth_client/lib/src/protocol/client.dart#L220

### How the endpoint method is identified on the server
The error message suggests that when the call reaches the server it contains the endpoint name (`adminAuth`?) but misses the method that is called there (`status`?).

The endpoint name is always supplies in the URI path while the method name is either supplied in the URI path or the POST request body.

The code for parsing it is defined here: https://github.com/serverpod/serverpod/blob/4af8a5c85954fe59fba00b00c1939215bdca6acb/packages/serverpod/lib/src/server/server.dart#L441-L460

### Suggestion for debugging
Do you have the possibility to inspect what the request looks like after it has passed through your nginx proxy? If so, does it look the sama as when you are doing requests locally?

**Comment by iwwencreative** [comment]
it occurs for all endpoints (in production)
all dependencies are up to date
i am running serverpod 2.1.5

here is the code for a call to custom endpoint i created for an admin to login (serverpod generate was run to update the client)

/front end

client.adminAuth
        .loginAdmin(
      emailController.text,
      passwordController.text,
    )
        .then((res) async {
      if (res.success) {
        await sessionManager.registerSignedInUser(
          res.userInfo,
          res.keyId.validate(),
          res.key.validate(),
        );
        setState(() {
          loading = false;
        });
        MainScreen().launch(context);
      } else {
        setState(() {
          loading = false;
          errorLoging = true;
        });
      }
    });
    
   PS: the call is made from a flutter web app 

i honestly dont know how to inspect the request after it has passed through nginx, any help on how to do that would be appreciated

**Comment by iwwencreative** [comment]
![Screenshot from 2024-11-27 14-15-16](https://github.com/user-attachments/assets/ae572856-81d0-42d5-94b4-1484b196ad79)
![Screenshot from 2024-11-27 14-15-39](https://github.com/user-attachments/assets/609caeea-131f-4205-84f0-83b254c6dcd8)

**Reply by HaroldRomero151199** [reply]
I have the same problem

**Comment by SandPod** [comment]
From the inspected content on the client, it seems like all data required to navigate to the correct endpoint is transmitted with the request.

My best guess is that the body of the request is stripped somewhere on its way to the Server.

Could you supply the configuration for your Nginx reverse proxy?

**Comment by iwwencreative** [comment]
server {
    listen 80;
    server_name my.ip.address; 

    location / {
        proxy_pass http://localhost:8080; 
        proxy_http_version 1.1;

        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "upgrade";

        # Standard headers
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Proto $scheme;

        proxy_set_header REMOTE-HOST $remote_addr;

        # Disable caching for WebSocket connections
        proxy_cache_bypass $http_upgrade;

        add_header 'Access-Control-Allow-Origin' '*';
        add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS';
        add_header 'Access-Control-Allow-Headers' 'Origin, Content-Type, X-Auth-Token, Authorization';
        add_header 'Access-Control-Allow-Credentials' 'true';

       
    }
}

**Reply by SandPod** [reply]
Unfortunately I'm not an nginx configuration expert, I think other forums other people in the community will be able to help out there. But here are two minor suggestions that can be tried out in isolation:

1. Add these two headers to ensure that the content type and content length are delivered to the server.

```
        proxy_set_header Content-Length $http_content_length;
        proxy_set_header Content-Type $http_content_type;
```

2. Disable request body buffering to see if this helps out.
```
        proxy_request_buffering off;
        proxy_buffering off;
```

**Comment by iwwencreative** [comment]
i have confirmed that nginx is messing with the request, i tried exposing my serverpod server directly and everything worked.
how do i fix nginx

**Comment by SandPod** [comment]
Converting this to a discussions since it is note related to an issue in Serverpod and updated the title to better reflect the issue.

**Comment by lodgmar** [comment]
I think the problem is because of browser sending `text/plain` content type, instead of `application/json`. Serverpod returns 400 error in release mode but not in develpment mode. Why is this happening? I think it's a bug in "serverpod_client_browser.dart", function `serverRequest()`.

**Reply by HaroldRomero151199** [reply]
how we solved the error

**Comment by iwwencreative** [comment]
i switched to traefik and everything worked

**Comment by HaroldRomero151199** [comment]
✅ [SOLVED] – Error 400 when calling Serverpod endpoints + WebSocket connection failure (Flutter Web + Nginx)

I ran into two issues when deploying a Serverpod backend behind Nginx, used with a Flutter Web frontend:

❌ POST requests were failing with 400 Bad Request: No method name specified in call to store
❌ WebSocket connections were throwing WebSocketChannelException

🛠 Fix: Updated my Nginx configuration

`# 🔁 Proxy para TODO tu backend (en 8080)
    location /api/ {
        proxy_pass http://nameContainer:8080/;
        #rewrite ^/api/(.*)$ /$1 break;

	# WebSocket support
	proxy_http_version 1.1;

	# Standard headers
	proxy_set_header Host $host;
	proxy_set_header X-Real-IP $remote_addr;
    	proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
   	proxy_set_header X-Forwarded-Proto $scheme;
	
	proxy_set_header REMOTE-HOST $remote_addr;

        # 🔐 CORS headers (si hacés llamadas desde Flutter Web u otros clientes)
        add_header 'Access-Control-Allow-Origin' '*';
        add_header 'Access-Control-Allow-Methods' 'GET, POST, PUT, DELETE, OPTIONS';
        add_header 'Access-Control-Allow-Headers' 'Origin, Content-Type, Accept, X-Auth-Token, Authorization, X-Requested-With' always;
        add_header 'Access-Control-Allow-Credentials' 'true';

        if ($request_method = 'OPTIONS') {
            add_header 'Content-Length' 0;
            add_header 'Content-Type' 'text/plain';
            return 204;
        }
    }

    location /v1/websocket {
    	proxy_pass http://nameContainer:8080;
	proxy_http_version 1.1;
	proxy_set_header Upgrade $http_upgrade;
	proxy_set_header Connection "upgrade";
	# Disable caching for WebSocket connections
    	proxy_cache_bypass $http_upgrade;
	# ...otros headers y configuración...
    }
}
`
📌 Results:

- POST requests from Flutter Web now work properly with application/json
- Serverpod’s WebSocket connection works without error
- No more 400 Bad Request or WebSocketChannelException

💬 Hopefully this helps someone else using Serverpod + Flutter Web behind Nginx!

**Comment by lodgmar** [comment]
I also finally made it work, with Nginx. This is my configuration for a Flutter web application with Serverpod, both client (www.mydomain.com) and server (www.mydomain.com/api) in docker images:

This goes into nginx.conf or some ".conf" file included early, before virtual severs (it has to be within "http" level). It's a Map used for websocket upgrading:
```
map $http_upgrade $connection_upgrade {
    default upgrade;
    ''      keep-alive;
}
```

And this is the virtual server configuration:
```
server {
    listen 443 ssl;
    server_name www.mydomain.com;
    access_log  /var/log/nginx/mydomain.com.access.log;
    error_log   /var/log/nginx/mydomain.com.error.log;

    # SSL
    ssl_certificate     /etc/ssl/mydomain.com.crt;
    ssl_certificate_key /etc/ssl/mydomain.com.key;
    include /etc/letsencrypt/options-ssl-nginx.conf;  # or whatever SSL configuration

    # Security headers for CORS
    add_header Strict-Transport-Security "max-age=63072000; includeSubdomains; preload" always;
    add_header X-Frame-Options "SAMEORIGIN" always;
    add_header X-Content-Type-Options "nosniff" always;
    add_header Referrer-Policy "strict-origin-when-cross-origin";

    location /api/ {
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto https;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_http_version 1.1;
        proxy_cache_bypass $http_upgrade;

        # Websockets: we use the map defined previously
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection $connection_upgrade; 

        client_max_body_size 1M;  # If you need to limit body size to 1M or something else

        proxy_pass http://serverpod-server-docker:8080/;   # change port to $SERVERPOD_API_SERVER_POD
    }

    location / {
        proxy_redirect off;
        proxy_set_header Host $host;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto https;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_http_version 1.1;
        proxy_cookie_flags ~ Secure;

        client_max_body_size 1M; # If you need to limit body size to 1M or something else

        proxy_pass http://serverpod-flutter-app-docker/;  
    }

}
```

This works for me, but I haven't used websocket yet, so maybe that part is wrong.
